app/models/miq_queue_worker_base/runner.rb

```diff
@@ -140,6 +167,36 @@ def drb_dequeue_available?
       end
   end
 
+  def miq_messaging_dequeue_available?
+    MiqQueue.messaging_type != "miq_queue" && MiqQueue.messaging_client(self.class.name).present?
+  end
+
+  def process_miq_messaging_message(_msg)
+    raise NotImplementedError, 'Must be implemented in subclass'
+  end
+
+  def miq_messaging_listener_thread
+    loop do
+      send("miq_messaging_subscribe_#{@worker.class.miq_messaging_subscribe_mode}") do |msg|
+        @message_queue << msg
+      end
+    rescue => err
+      _log.warn("miq_messaging_listener_thread error [#{err}]")
+    end
+  end
+
+  def miq_messaging_subscribe_topic(&block)
+    messaging_client = MiqQueue.messaging_client(self.class.name)
```

Comment:
I notice this bit duplicated in a number of places - would it make sense to move this out to a `messaging_client` method?